capstone/src/main/java/com/google/sps/servlets/profile/ProfileServlet.java

```diff
@@ -0,0 +1,113 @@
+    this.userService = userService;
+    this.datastore = datastore;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Obtain userId from param URL.
+    String[] idArray = request.getPathInfo().split("/");
+    if (idArray.length < 2) {
+      response.sendError(
+          HttpServletResponse.SC_NOT_FOUND,
+          "The profile you were looking for was not found in our records!");
+      return;
+    }
+
+    String userId = idArray[1];
+
+    Key userKey = KeyFactory.createKey("UserProfile", userId);
+    Entity entity;
+
+    try {
+      entity = datastore.get(userKey);
+    } catch (EntityNotFoundException e) {
+      System.err.println("Could not find key: " + userKey);
+      response.sendError(
+          HttpServletResponse.SC_NOT_FOUND,
+          "The profile you were looking for was not found in our records!");
+      return;
+    }
+
+    // If userId is a business owner id, redirect to "profile not found" page.
+    String isBusiness =
+        entity.hasProperty("isBusiness") ? (String) entity.getProperty("isBusiness") : "";
+    if (isBusiness.equals("Yes")) {
+      response.sendError(
+          HttpServletResponse.SC_NOT_FOUND,
+          "The profile you were looking for was not found in our records!");
+      return;
+    }
+
+    // Query all profile properties.
+    String id = entity.getKey().getName();
+    String name = entity.hasProperty("name") ? (String) entity.getProperty("name") : "Anonymous";
+    String location = entity.hasProperty("location") ? (String) entity.getProperty("location") : "";
+    String bio = entity.hasProperty("bio") ? (String) entity.getProperty("bio") : "";
+    boolean isCurrentUser = userId.equals(id);
+
+    // Create a profile object that contains the properties.
+    UserProfile profile = new UserProfile(id, name, location, bio, isCurrentUser);
+
+    // Send it back to client side as a JSON file.
+    response.setContentType("application/json;");
+    Gson gson = new Gson();
+    response.getWriter().println(gson.toJson(profile));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // When user edits their profile page, fetch the data from the edit profile form.
+    String id = userService.getCurrentUser().getUserId();
+    String isBusiness = request.getParameter("isBusiness");
+    String name = request.getParameter("name");
+    String location = request.getParameter("location");
+    String bio = request.getParameter("bio");
+
+    // Check if user is logged in.
+    if (id == null) {
+      response.sendError(
+          HttpServletResponse.SC_NOT_FOUND, "You don't have permission to perform this action!");
+      return;
+    }
+    // Update properties in datastore.
+    Entity profileEntity = new Entity("UserProfile", id);
+    profileEntity.setProperty("isBusiness", isBusiness);
```

Comment:
```suggestion
    profileEntity.setProperty("isBusiness", request.getParameter("isBusiness"));
```
We can do that for each of these :)

Reply:
done